design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+* Assets will not be supported in environment-agnostic stacks. [#4131](https://github.com/aws/aws-cdk/pull/4131) proposes that `cdk deploy` will default `env` to the current account/region, which means that the CLI use case will no longer treat stacks as environment-agnostic.
+
+## Approach
+
+At a high-level, we will model the deployment process of a CDK app as follows:
+
+```
+bootstrap => source => build => synthesis => mutate => publish => deploy
+```
+
+1. **bootstrap**: manually pre-provision resources required to deploy CDK apps into this environment (such as an S3 bucket, ECR repository and various IAM roles that trust the central deployment account).
+2. **source**: the code is pulled from a source repository (e.g. CodeCommit, GitHub or S3), like any other app.
+3. **build**: compiles the CDK app code into an executable program (user-defined).
+4. **synthesis**: invokes the compiled executable to produce a [cloud assembly](https://github.com/aws/aws-cdk/blob/master/design/cloud-assembly.md) from the app. It includes a CloudFormation template for each stack and asset sources (docker images, s3 files, etc) that must be packaged and published to the asset store in each environment that consumes them.
+5. **mutate**: update stack(s) required by the pipeline. This includes pipeline resources and other auxiliary resources such as regional replication buckets. These stacks are limited to 50KiB and are not allowed to use assets, so they can be deployed without bootstrapping resources.
+6. **publish**: package and publish all assets to asset stores (S3 bucket, ECR repository) so they can be consumed.
+7. **deploy**: stage(s), stacks are deployed to the various environments through some orchestration process (e.g. deploy first to this region, run these canaries, wait for errors, continue to the next stage, etc).
+
+NOTE: The deployment phase can include any number of stack deployment actions. Each deployment action is responsible deploy a single stack, along with any assets it references.
+
+This following sections describes the design of each component in the toolchain.
+
+## Build
+
+In this stage we compile the CDK app to an executable program through a user-defined build system.
+
+We assume a single source repository which masters the CDK app itself. This repo can be structured in any way users wish, and may include multiple modules or just a single one. If users choose to organize their project into modules and master different modules in other repositories, eventually the build artifacts from these builds should be available when the app is synthesized.
+
+The only requirement from the build step is the it will have an output artifact that is a cloud-assembly directory which is obtained through `cdk synth` (defaults to `./cdk.out`). Other than that, users can fully control their build environment.
```

Comment:
Isn't this one of the outputs of the next step, rather - synthesis? Are you suggesting the package build create `cdk.out/`?

Isn't there another requirement of the this step is the existence of the `cdk.json` with some required parameters? Said another way, the `cdk synth` command should work.

Reply:
Merged build + synthesis